### DISCUSSION TITLE
Mermaid support out of the box

### DISCUSSION THREAD

**Discussion by TheKnarf** [withastro/starlight#1259] [Feature Requests]
### What version of `starlight` are you using?

N/A

### What is your idea?

Starlight should support Mermaid diagrams either our of the box, or as an official plugin. It's really dumb that it doesn't

### Why is this feature necessary?

Setting it up by yourself is a ton of work, and its really buggy.

You can do so by adding the following to `astro.config.mjs`:

```javascript
import rehypeMermaid from 'rehype-mermaid';
import rehypeRaw from 'rehype-raw';
//...
	markdown: {
		rehypePlugins: [
			rehypeRaw,
			rehypeMermaid,
		],
	},
```

However if you do so you MDX support crashes. So you also have to add the following:

```javascript
export default defineConfig({
  integrations: [
    starlight({
    // ...
    }),
   astroExpressiveCode(),
   mdx({
      // Markdown config now ignored
      extendMarkdownConfig: false,
      // No `remarkPlugins` applied
    }),
   ],
})
```

And after all of that work it only works when `starlight` is in light-mode, as it only generate styling for one mode out of the box. Adding support to integrate with `starlight`s dark- and light-mode would be even more work.

This should ultimatly just be supported by `startlight` if they want to be serious alternative.

### Do you have examples of this feature in other projects?

Docusaurus have an official plugin thats super easy to use

https://docusaurus.io/docs/markdown-features/diagrams

### Participation

- [ ] I am willing to submit a pull request for this feature.

**Comment by stereobooster** [comment]
**UPD** see solution in the comment bellow https://github.com/withastro/starlight/discussions/1259#discussioncomment-9300105

Article about different configuration options https://astro-digital-garden.stereobooster.com/recipes/mermaid-diagrams-in-markdown/

<details>
  <summary>old answer</summary>
I think I got to the bottom of the problem. There are two options:

## remark-mermaidjs

Use [`remark-mermaidjs`](https://github.com/remcohaszing/remark-mermaidjs), so it would process code blocks before [`remark-shiki`](https://github.com/withastro/astro/blob/main/packages/markdown/remark/src/remark-shiki.ts)

**But** author (of the plugin) [advice against it](https://github.com/remcohaszing/remark-mermaidjs/issues/23#issuecomment-1881313556)

## rehype-mermaidjs

It would work with `remark-shiki` unless it would mess with `<pre>`/`<code>` tags.

In this case you need either:

- use `rehype-raw`, like in the original post
- ask Astro team to fix `remark-shiki`
- disable `remark-shiki` at all ([`markdown.syntaxHighlight: false`](https://docs.astro.build/en/reference/configuration-reference/#markdownsyntaxhighlight)) and use [`rehype-shikiji`](https://shikiji.netlify.app/packages/rehype)

## PS

My post about [Astro diagram solutions](https://stereobooster.com/posts/astro-diagrams/)
</details>

**Comment by icelattefull** [comment]
None of the approaches listed above worked for me. However, [this](https://github.com/withastro/astro/issues/4433#issuecomment-1584019991) did

**Comment by delucis** [comment]
Not Mermaid, but sharing an Astro integration for the D2 diagramming language that @HiDeoo built that might also be of interest to people in this thread: https://astro-d2.vercel.app/

**Reply by stereobooster** [reply]
`d2` uses `playwright` (and `goja`) the same as `remark-mermaidjs`. Funny that whatever we do - we get back to headless browsers

**Reply by HiDeoo** [reply]
To recap everything I learned while working on this:

- `d2` *only* uses Playwright for PNG exports ([reference](https://d2lang.com/tour/exports/#png))
- `astro-d2` never exports to PNG
- `remark-mermaidjs` uses Playwright for all exports and is not recommended by its author who recommends using `rehype-mermaid` instead ([reference](https://github.com/remcohaszing/remark-mermaidjs))
- `rehype-mermaid` uses Playwright for all exports

**Reply by delucis** [reply]
I’m pretty sure PNG support could be added without Playwright too if desired using a tool like https://github.com/yisibl/resvg-js

**Reply by HiDeoo** [reply]
> I’m pretty sure PNG support could be added without Playwright too if desired using a tool like [yisibl/resvg-js](https://github.com/yisibl/resvg-js)

Definitely, this is even tracked by D2 but currently blocked by https://github.com/RazrFalcon/resvg/issues/541

**Reply by stereobooster** [reply]
Thank you, @HiDeoo for pointing this out. Somehow I was sure it requires Playwright for everything. I will need to play with it more

**Comment by darrenparkinson** [comment]
Also interested in this capability. Pretty much makes it unusable for tech docs without it. We're currently testing with [mdbook](https://rust-lang.github.io/mdBook/) which [you can add mermaid to](https://github.com/badboy/mdbook-mermaid) very easily. 

Docs currently built with Docker using this:

```dockerfile
FROM rust as book-build
RUN cargo install mdbook
RUN cargo install mdbook-mermaid
WORKDIR /app
COPY --from=preparation-stage /app .
RUN mdbook build
```

We're not a rust shop though, so was looking at alternatives. 

mkdocs also appears to support mermaid fairly easily too.

This would genuinely be a great addition out of the box.

**Comment by delucis** [comment]
**EDIT:** This answer is outdated. See @HiDeoo’s [Mermaid solution below](https://github.com/withastro/starlight/discussions/1259#discussioncomment-9300105) for the latest and simplest set-up.

<details>
<summary>Original comment</summary>

Just did some testing to get [`reype-mermaid`](https://github.com/remcohaszing/rehype-mermaid) working kind of OK and sharing notes here in case it’s helpful

1. A few dependencies are needed, mostly for massaging the remark/rehype pipeline to work nicely:

	```sh
	npm i hast-util-from-html hast-util-to-html mdast-util-to-hast rehype-mermaid unist-util-visit
	```

2. Add this module to your project. It exports an Astro integration that wires up `rehype-mermaid` with some small utilities to ensure code blocks are not transformed too early.

	```js
	// integrations/astro-rehype-mermaid.js
	
	import { fromHtml } from 'hast-util-from-html';
	import { toHtml } from 'hast-util-to-html';
	import { toHast } from 'mdast-util-to-hast';
	import { visit } from 'unist-util-visit';
	import rehypeMermaid from 'rehype-mermaid';
	
	/** Transform a mermaid code block to HTML before it gets syntax highlighted to preserve its structure. */
	const remarkMermaidToHtml = () => (tree) => {
	  visit(tree, 'code', (code, index, parent) => {
	    if (index === undefined || !parent) return;
	    if (code.lang === 'mermaid') {
	      const html = toHtml(toHast(code));
	      parent.children.splice(index, 1, { type: 'html', value: html });
	    }
	  });
	};
	
	/** Transform a raw HTML node created from a mermaid code block back to a full HAST. */
	const rehypeMermaidRawToHast = () => (tree) => {
	  visit(tree, 'raw', (node, index, parent) => {
	    if (index === undefined || !parent) return;
	    if (node.value.startsWith('<pre><code class="language-mermaid"')) {
	      const hast = fromHtml(node.value, { fragment: true });
	      parent.children.splice(index, 1, hast);
	    }
	  });
	};
	
	/**
	 * Astro integration to set up `rehype-mermaid` to render Markdown code blocks with the `mermaid` language.
	 * @param {import('rehype-mermaid').RehypeMermaidOptions} options
	 * @return {import('astro').AstroIntegration}
	 */
	export default function astroRehypeMermaid(options = {}) {
	  return {
	    name: 'astro-mermaid',
	    hooks: {
	      'astro:config:setup'({ updateConfig }) {
	        updateConfig({
	          markdown: {
	            remarkPlugins: [remarkMermaidToHtml],
	            rehypePlugins: [rehypeMermaidRawToHast, [rehypeMermaid, options]],
	          },
	        });
	      },
	    },
	  };
	}
	```

3. Import the `astroRehypeMermaid` integration into `astro.config.mjs` and add it to your `integrations` array:

	```js
	// astro.config.mjs
	import { defineConfig } from 'astro/config';
	import starlight from '@astrojs/starlight';
	import astroRehypeMermaid from './integrations/astro-rehype-mermaid';
	
	export default defineConfig({
	  integrations: [
	    astroRehypeMermaid({
	      // rehype-mermaid options can be passed here
	    }),
	    starlight({
	      title: 'My Docs',
	    }),
	  ],
	});
	```

I didn’t get far enough to work out if there’s a nicer way to handle dark/light mode support, but maybe this is a good starting point for someone to iterate on or even publish as a community package!

</details>

**Reply by stereobooster** [reply]
<details>
  <summary>old answer</summary>

~~Hey, @delucis just out of curiosity why Astro default highlighter is at remark level instead of rehype? Is there some technical reason for it or it's just historically this way?~~

~~For example, `@shikijs/rehype` works in Astro. And this is the same highlighter that Astro uses, but at rehype level.~~

I was wrong with my assumption. Astro uses rehype for code highlighting. The problem is that it loads highlighter before any other rehype plugins:

```js
	if (!isPerformanceBenchmark) {
		// Syntax highlighting
		if (syntaxHighlight === 'shiki') {
			parser.use(rehypeShiki, shikiConfig);
		} else if (syntaxHighlight === 'prism') {
			parser.use(rehypePrism);
		}
	}

	// User rehype plugins
	for (const [plugin, pluginOpts] of loadedRehypePlugins) {
		parser.use(plugin, pluginOpts);
	}
```

For Astro solution can be:

```astro.config.mjs
import { defineConfig } from "astro/config";
import { rehypeShiki, markdownConfigDefaults } from "@astrojs/markdown-remark";
import rehypeMermaid from "rehype-mermaid";

// https://astro.build/config
export default defineConfig({
  markdown: {
    syntaxHighlight: false,
    rehypePlugins: [
      rehypeMermaid,
      [rehypeShiki, markdownConfigDefaults.shikiConfig],
    ],
  },
});
```

Need to check if it will work for Starlight though
</details>

**Reply by stereobooster** [reply]
<details>
  <summary>old answer</summary>

Here is very hacky way (but **the point is that rehypeMermaid works out of the box** if it would be possible to put it before `rehypeExpressiveCode`):

```js
import { defineConfig } from "astro/config";
import starlight from "@astrojs/starlight";
import rehypeMermaid from "rehype-mermaid";
import { astroExpressiveCode, addClassName } from "astro-expressive-code";

export default defineConfig({
  integrations: [
    // astroExpressiveCode inserted before starlight it inserts `rehypeExpressiveCode` after `rehypeMermaid` and it works
    astroExpressiveCode({ ... }),
    starlight({ ... }),
  ],
  markdown: {
    rehypePlugins: [rehypeMermaid],
  }
})
```

**Can this be reported as bug**?


How to configure `astroExpressiveCode` so it would look more like default Starlight style: 

```js
 astroExpressiveCode({
  customConfigPreprocessors: {
    preprocessAstroIntegrationConfig: () => {
      // Add the `not-content` class to all rendered blocks to prevent them from being affected
      // by Starlight's default content styles
      const plugins = [
        {
          name: "Starlight Plugin",
          hooks: {
            postprocessRenderedBlock: ({ renderData }) => {
              addClassName(renderData.blockAst, "not-content");
            },
          },
        },
      ];

      return {
        themeCssSelector: (theme, { styleVariants }) => {
          // If one dark and one light theme are available, and the user has not disabled it,
          // generate theme CSS selectors compatible with Starlight's dark mode switch
          if (styleVariants.length >= 2) {
            const baseTheme = styleVariants[0]?.theme;
            const altTheme = styleVariants.find(
              (v) => v.theme.type !== baseTheme?.type
            )?.theme;
            if (theme === baseTheme || theme === altTheme)
              return `[data-theme='${theme.type}']`;
          }
          // Return the default selector
          return `[data-theme='${theme.name}']`;
        },
        plugins,
      };
    },
  },
});
```
</details>

**Reply by delucis** [reply]
In our next release, we’ll be updating Expressive Code — see https://github.com/withastro/starlight/pull/1720 — not 100% sure, but it also moves stuff to rehype so might help.

**Reply by HiDeoo** [reply]
Starting with Starlight [v0.22](https://github.com/withastro/starlight/releases/tag/%40astrojs%2Fstarlight%400.22.0), which includes an [update to Expressive Code](https://expressive-code.com/releases/#0350) transitioning from a remark plugin to a rehype plugin, the instructions shared by Chris above can be simplified a bit.

1. Install `rehype-mermaid`

   ```sh
   npm i rehype-mermaid
   ```

2. Add the Mermaid rehype plugin to `astro.config.mjs`:

   ```js
   // astro.config.mjs
   import { defineConfig } from "astro/config";
   import starlight from "@astrojs/starlight";
   import rehypeMermaid from "rehype-mermaid";

   export default defineConfig({
     integrations: [
       starlight({
         title: "My Docs",
       }),
     ],
     markdown: {
       rehypePlugins: [rehypeMermaid],
     },
   });
   ```

**Reply by stereobooster** [reply]
Thank you very much for fixing it.

```js
export default defineConfig({
  markdown: {
    rehypePlugins: [[rehypeMermaid, { strategy: "img-svg", dark: true }]],
  },
});
```

Does anybody know how to solve issue with dark mode? https://github.com/withastro/starlight/discussions/1829

**Reply by TeenBiscuits** [reply]
I have a problem with this implementation and it is the dependency on playwright. If you try to deploy your website to vercel there is no way to use playwright because you can't install system packages.

<details><summary>Details Vercel Logs</summary>
<p>

```shell
╔══════════════════════════════════════════════════════╗
║ Host system is missing dependencies to run browsers. ║
║ Please install them with the following command:      ║
║                                                      ║
║     pnpm exec playwright install-deps                ║
║                                                      ║
║ Alternatively, use apt:                              ║
║     apt-get install libnss3\                         ║
║         libnspr4\                                    ║
║         libgbm1                                      ║
║                                                      ║
║ <3 Playwright Team                                   ║
╚══════════════════════════════════════════════════════╝
```

</p>
</details> 

I use this implementation [Adding Mermaid diagrams to Astro MDX - Xkonti.tech](https://xkonti.tech/blog/astro-mermaid-mdx/). If there is someone who knows how to use this package in vercel I would prefer to be able to use a rehype plugin.

By the way i use vercel SSR.

**Reply by HuakunShen** [reply]
@stereobooster how to make the theme reactive? Support both dark and light theme based on current page preference.

**Reply by stereobooster** [reply]
https://astro-digital-garden.stereobooster.com/recipes/mermaid-diagrams-in-markdown/ `strategy: "img-class-dark-mode"`

**Reply by asbjornu** [reply]
I got @HiDeoo's recipe in https://github.com/withastro/starlight/discussions/1259#discussioncomment-9300105 to work after upgrading Astro from v4.2.8 to v4.16.18 and Starlight from v0.15.4 to v0.29.3.

**Comment by torn4dom4n** [comment]
I found [this site](https://onerepo.tools/concepts/primer/#resolving-dependencies) which is so good to support Mermaid.

**Reply by stereobooster** [reply]
<details>
  <summary>old answer</summary>

here is what they use https://github.com/paularmstrong/onerepo/blob/main/docs/src/plugins/remark-mermaid.ts

```ts
import type { RemarkPlugin } from '@astrojs/markdown-remark';
import { visit } from 'unist-util-visit';

const escapeMap: Record<string, string> = {
	'&': '&amp;',
	'<': '&lt;',
	'>': '&gt;',
	'"': '&quot;',
	"'": '&#39;',
};

const escapeHtml = (str: string) => str.replace(/[&<>"']/g, (c) => escapeMap[c]);

export const mermaid: RemarkPlugin<[]> = () => (tree) => {
	visit(tree, 'code', (node) => {
		if (node.lang !== 'mermaid') return;

		// @ts-ignore
		node.type = 'html';
		node.value = `<div class="mermaid" data-content="${escapeHtml(node.value)}"><p>Loading graph...</p></div>`;
	});
};
```
</details>

**Reply by torn4dom4n** [reply]
Yeah, I know it. But I still hope we have an integratetion or component to support this.

**Comment by stereobooster** [comment]
Related: implemented small library to add pan/zoom functionality to SVG, img - https://github.com/stereobooster/svg-pan-zoom-gesture

**Comment by stereobooster** [comment]
Just in case if anybody is interested, `rehype-raw` that is used by Astro may cause issues for some diagrams. For example, https://github.com/remcohaszing/rehype-mermaid/issues/17

**Comment by ematipico** [comment]
Now that Astro allows us to exclude languages, I think it's possible to integrate the support out of the box with little changes: https://docs.astro.build/en/reference/configuration-reference/#markdownsyntaxhighlightexcludelangs

**Reply by delucis** [reply]
I don’t think we were impacted that because Expressive Code already handled it and we’re not using Astro’s Shiki instance. The main reason not to support it out of the box is just that Mermaid is a large, slow, somewhat old-fashioned package, so bundling it with Starlight is probably not a great idea. A third-party plugin for Starlight would probably be the best approach to allow people to add Mermaid support when they need it.

**Comment by awhitford** [comment]
> Setting it up by yourself is a ton of work

This was a driver for creating Starlight Starter. Starlight is great, but it is time consuming to configure a new project with the support that you need like Mermaid diagrams, Blog, build scripts, etc.  So this approach is simple:  you fill out a form, check the options that you want, and download your project -- admittedly, for a modest fee.

Some information on what the project does to enable Mermaid is available:  [Enabling Mermaid](https://www.starlightstarter.com/documentation/options/mermaid/#enabling-mermaid).

**Reply by HiDeoo** [reply]
Please refrain from posting advertisements for your paid services in this GitHub repository. This is an open-source project community, and we encourage collaboration and sharing of knowledge without any commercial intent. Thank you for your understanding!

**Comment by pppwaw** [comment]
I found a repo [astro-mermaid](https://github.com/joesaby/astro-mermaid) which provides client-side Mermaid support. It works well for me!
